float32[2] pos
float32[2] dir
float32 angle
float32[2] upper_b_s
float32[2] lower_b_s
float32[2] upper_b_e
float32[2] lower_b_e
